README.md
# A powerful, easy to configure uptime monitor

[![Latest Version on Packagist](https://img.shields.io/packagist/v/spatie/laravel-uptime-monitor.svg?style=flat-square)](https://packagist.org/packages/spatie/laravel-uptime-monitor)
[![MIT Licensed](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/spatie/laravel-uptime-monitor/master.svg?style=flat-square)](https://travis-ci.org/spatie/laravel-uptime-monitor)
[![Quality Score](https://img.shields.io/scrutinizer/g/spatie/laravel-uptime-monitor.svg?style=flat-square)](https://scrutinizer-ci.com/g/spatie/laravel-uptime-monitor)
[![StyleCI](https://styleci.io/repos/67774357/shield?branch=master)](https://styleci.io/repos/67774357)
[![Total Downloads](https://img.shields.io/packagist/dt/spatie/laravel-uptime-monitor.svg?style=flat-square)](https://packagist.org/packages/spatie/laravel-uptime-monitor)

Laravel-uptime-monitor is a powerful, easy to configure uptime monitor. It will notify you when your site is down (and when it comes back up). You can also be notified a few days before an SSL certificate on one of your sites expires. Under the hood, the package leverages Laravel native notifications, so it's easy to use Slack, Telegram or your preferred notification provider.

You'll find extensive documentation on [https://docs.spatie.be/laravel-uptime-monitor]. It includes detailed info on how to install and use the package.

If you're not familiar with Laravel, but still want to use our uptime monitor, take a look at the [uptime-monitor-app](https://github.com/spatie/uptime-monitor-app) repo which contains a stand alone version of this package.

Reading the config file of this package is a good way to quickly get a feel of what `laravel-uptime-monitor` can do. Here's the content of the config file:

```php
return [

    /*
     * You can get notified when specific events occur. Out of the box you can use 'mail'
     * and 'slack'. Of course you can also specify your own notification classes.
     */
    'notifications' => [

        'notifications' => [
            \Spatie\UptimeMonitor\Notifications\Notifications\UptimeCheckFailed::class => ['slack'],
            \Spatie\UptimeMonitor\Notifications\Notifications\UptimeCheckRecovered::class => ['slack'],
            \Spatie\UptimeMonitor\Notifications\Notifications\UptimeCheckSucceeded::class => [],

            \Spatie\UptimeMonitor\Notifications\Notifications\CertificateCheckFailed::class => ['slack'],
            \Spatie\UptimeMonitor\Notifications\Notifications\CertificateExpiresSoon::class => ['slack'],
            \Spatie\UptimeMonitor\Notifications\Notifications\CertificateCheckSucceeded::class => [],
        ],

        /*
         * The location from where you are running this Laravel application. This location will be 
         * mentioned in all notifications that will be sent.
         */
        'location' => '',

        /*
         * To keep reminding you that an uptime check failed, notifications
         * will be resent every given number of minutes.
         */
        'resend_uptime_check_failed_notification_every_minutes' => 60,

        'mail' => [
            'to' => 'your@email.com',
        ],

        'slack' => [
            'webhook_url' => env('UPTIME_MONITOR_SLACK_WEBHOOK_URL'),
        ],

        /*
         * Here you can specify the notifiable to which the notifications should be sent. The default
         * notifiable will use the variables specified in this config file.
         */
        'notifiable' => \Spatie\UptimeMonitor\Notifications\Notifiable::class,
        
        /**
         * The date format used in notifications.
         */
        'date_format' => 'd/m/Y',
    ],

    'uptime_check' => [
        /*
         * When the uptime check could reach the url of a monitor it will pass the response to this class
         * If this class determines the response is valid, the uptime check will be regarded as succeeded.
         *
         * You can use any implementation of Spatie\UptimeMonitor\Helpers\UptimeResponseCheckers\UptimeResponseChecker here.
         */
        'response_checker' => Spatie\UptimeMonitor\Helpers\UptimeResponseCheckers\LookForStringChecker::class,
            
        /*
         * An uptime check will be performed if the last check was performed more than the
         * given number of minutes ago. If you change this setting you have to manually
         * update the `uptime_check_interval_in_minutes` value of your existing monitors.
         *
         * When an the uptime check failed we'll check the uptime every time `monitor:check-uptime` runs
         * regardless of this setting.
         */
        'run_interval_in_minutes' => 5,

        /*
         * To speed up the uptime checking process uptime monitor can perform the uptime check of several
         * monitors concurrently. Set this to a lower value if you're getting weird errors
         * running the uptime check.
         */
        'concurrent_checks' => 10,

        /*
         * The uptime check will fail if the configured url does not respond after the
         * given number of seconds.
         */
        'timeout_per_site' => 10,

        /*
         * Fire `Spatie\UptimeMonitor\Events\MonitorFailed` event only after
         * the given number of uptime checks have consecutively failed for a monitor.
         */
        'fire_monitor_failed_event_after_consecutive_failures' => 2,

        /*
         * When reaching out to sites this user agent will be used.
         */
        'user_agent' => 'spatie/laravel-uptime-monitor uptime checker',
        
        /*
         * When reaching out to the sites these headers will be added.
         */
        'additional_headers' => []
    ],

    'certificate_check' => [

        /*
         * The `Spatie\UptimeMonitor\Events\SslExpiresSoon` event will fire
         * when a certificate is found whose expiration date is in
         * the next number of given days.
         */
        'fire_expiring_soon_event_if_certificate_expires_within_days' => 10,
    ],

    'expiration_check' => [

        /*
         * The `Spatie\UptimeMonitor\Events\DomainExpiresSoon` event will fire
         * when a domain is found whose expiration date is in
         * the next number of given days.
         */
        'fire_expiring_soon_event_if_domain_expires_within_days' => 30,

        /**
         * @todo Config option for custom whois servers
         * @see https://packagist.org/packages/io-developer/php-whois
         * @see https://github.com/io-developer/php-whois
         */
        'whois_servers' => [],
    ],

    /*
     * To add or modify behaviour to the Monitor model you can specify your
     * own model here. The only requirement is that it should extend
     * `Spatie\UptimeMonitor\Models\Monitor`.
     */
     'monitor_model' => Spatie\UptimeMonitor\Models\Monitor::class,
];
```

## API

Currently this package does not offer an API, if you need that take a look at [this package](https://github.com/LKDevelopment/laravel-uptime-monitor-api).

## Documentation
You'll find the documentation on [https://docs.spatie.be/laravel-uptime-monitor]. It includes detailed info on how to install and use the package.

Find yourself stuck using the package? Found a bug? Do you have general questions or suggestions for improving the uptime monitor? Feel free to [create an issue on GitHub](https://github.com/spatie/laravel-uptime-monitor/issues), we'll try to address it as soon as possible.

## Changes with this Fork

### Manage Monitors by ID
Added ability to view and use the ID of the row to manage monitors. E.g. the list command will show all with the ID field. I find it easier to manage with IDs than copying / pasting the full urls.
```
php artisan monitor:list
```

Then if you need to disable/enable/check-uptime on a monitor with a given ID or 
```
php artisan monitor:disable 42
php artisan monitor:enable 42
php artisan monitor:check-uptime 42
```

### New Monitor - Check Domain Registration
Added a whois lookup on the domain to monitor if a domain is going to expire within the configured window. Config option:
```
'fire_expiring_soon_event_if_domain_expires_within_days' => 30,
```

New monitor examples:
```
monitor:check-expiration
monitor:check-expiration --url=https://mardala.com
monitor:check-expiration --id=42
```

#### TODO
* Create tests for the domain expiration monitor
* Create tests for using IDs
* Create method for modifying / updating the data schema if using an existing installation.
* Create a stand alone package with the entire install

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

To run the tests, issue this command:

```bash
vendor/bin/phpunit
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email freek@spatie.be instead of using the issue tracker.

## Postcardware

You're free to use this package, but if it makes it to your production environment we highly appreciate you sending us a postcard from your hometown, mentioning which of our package(s) you are using.

Our address is: Spatie, Samberstraat 69D, 2060 Antwerp, Belgium.

We publish all received postcards [on our company website](https://spatie.be/en/opensource/postcards).

## Credits

- [Freek Van der Herten](https://github.com/freekmurze)
- [All Contributors](../../contributors)

## Support us

Spatie is a webdesign agency based in Antwerp, Belgium. You'll find an overview of all our open source projects [on our website](https://spatie.be/opensource).

Does your business depend on our contributions? Reach out and support us on [Patreon](https://www.patreon.com/spatie). 
All pledges will be dedicated to allocating workforce on maintenance and new awesome stuff.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[https://docs.spatie.be/laravel-uptime-monitor]: https://docs.spatie.be/laravel-uptime-monitor/v3


CHANGELOG.md
# Changelog

All notable changes to `laravel-uptime-monitor` will be documented in this file

## Unreleased
### Added
- Ability to use IDs instead of URLs to manage monitors
- Added new domain expiration checking feature

## 2.6.9 - 2019-09-10
### Added
- Compatibility with Laravel 6.0 [#188](https://github.com/spatie/laravel-uptime-monitor/pull/188)

## 3.5.0 - 2019-05-17
### Added
- Add `raw_url` attribute to serialization [#175](https://github.com/spatie/laravel-uptime-monitor/pull/175)

## 3.4.1 - 2019-04-15
### Fixed
- Fixed issue with migrations stub ([#171](https://github.com/spatie/laravel-uptime-monitor/pull/171))

## 3.4.0 - 2019-03-03
### Changed
* Dropped support for Laravel 5.7
* Added support for Laravel 5.8, PHPUnit 8
* PHPUnit minimum version is now 7.5

## 3.3.4 - 2018-10-30

- fix if statement to be if not certificate applies to url

## 3.3.3 - 2018-10-20

- fix for PHP 7.3

## 3.3.2 - 2018-10-18

- fix for checking SSL

## 3.3.1 - 2018-08-27

- add support for Laravel 5.7

## 3.3.0 - 2018-03-13

- add option to force run all monitors

## 3.2.1 - 2018-02-08

- add support for L5.6

## 3.2.0 - 2017-12-20

- add ability for monitors to have their own response checkers

## 3.1.0 - 2017-12-11

- add ability to send payload to verify uptime

## 3.0.0 - 2017-08-31

- add support for Laravel 5.5, drop support for Laravel 5.4
- renamed config file from `laravel-uptime-monitor` to `uptime-monitor`

## 2.2.0 - 2017-03-13

- add `retry_connection_after_milliseconds` to config file

## 2.1.0 - 2017-03-13

- add `sync` command

## 2.0.3 - 2017-03-13

- fixed bug in getting unchecked monitors

## 2.0.2 - 2017-03-08

- added monitor location to mail notifications

## 2.0.1 - 2017-01-27

- ask for protocol when creating a monitor

## 2.0.0 - 2017-01-24

- add support for L5.4
- drop support for L5.3

## 1.2.3 - 2017-01-14

- fixed bug where migration could be published multiple times

## 1.2.2 - 2017-01-06

- set fallback text for Slack notifications

## 1.2.1 - 2016-12-22

- fix typos in notifications

## 1.2.0 - 2016-12-22

- improve notifications

## 1.1.2 - 2016-12-19

- fix `CertificateCheckSucceeded` notification

## 1.1.1 - 2016-12-12

- fix typos in command descriptions

## 1.1.0 - 2016-12-03

- added `additional_headers` to config

## 1.0.2 - 2016-11-24

- fix descriptions in config file

## 1.0.1 - 2016-11-21

- fix custom model instructions in config file

## 1.0.0 - 2016-11-21

- initial release


config/uptime-monitor.php
<?php

return [

    /*
     * You can get notified when specific events occur. Out of the box you can use 'mail'
     * and 'slack'. Of course you can also specify your own notification classes.
     */
    'notifications' => [

        'notifications' => [
            \Spatie\UptimeMonitor\Notifications\Notifications\UptimeCheckFailed::class => ['slack'],
            \Spatie\UptimeMonitor\Notifications\Notifications\UptimeCheckRecovered::class => ['slack'],
            \Spatie\UptimeMonitor\Notifications\Notifications\UptimeCheckSucceeded::class => [],

            \Spatie\UptimeMonitor\Notifications\Notifications\CertificateCheckFailed::class => ['slack'],
            \Spatie\UptimeMonitor\Notifications\Notifications\CertificateExpiresSoon::class => ['slack'],
            \Spatie\UptimeMonitor\Notifications\Notifications\CertificateCheckSucceeded::class => [],
        ],

        /*
         * The location from where you are running this Laravel application. This location will be
         * mentioned in all notifications that will be sent.
         */
        'location' => '',

        /*
         * To keep reminding you that a site is down, notifications
         * will be resent every given number of minutes.
         */
        'resend_uptime_check_failed_notification_every_minutes' => 60,

        'mail' => [
            'to' => ['your@email.com'],
        ],

        'slack' => [
            'webhook_url' => env('UPTIME_MONITOR_SLACK_WEBHOOK_URL'),
        ],

        /*
         * Here you can specify the notifiable to which the notifications should be sent. The default
         * notifiable will use the variables specified in this config file.
         */
        'notifiable' => \Spatie\UptimeMonitor\Notifications\Notifiable::class,

        /*
         * The date format used in notifications.
         */
        'date_format' => 'd/m/Y',
    ],

    'uptime_check' => [

        /*
         * When the uptime check could reach the url of a monitor it will pass the response to this class
         * If this class determines the response is valid, the uptime check will be regarded as succeeded.
         *
         * You can use any implementation of Spatie\UptimeMonitor\Helpers\UptimeResponseCheckers\UptimeResponseChecker here.
         */
        'response_checker' => Spatie\UptimeMonitor\Helpers\UptimeResponseCheckers\LookForStringChecker::class,

        /*
         * An uptime check will be performed if the last check was performed more than the
         * given number of minutes ago. If you change this setting you have to manually
         * update the `uptime_check_interval_in_minutes` value of your existing monitors.
         *
         * When an uptime check fails we'll check the uptime for that monitor every time `monitor:check-uptime`
         * runs regardless of this setting.
         */
        'run_interval_in_minutes' => 5,

        /*
         * To speed up the uptime checking process the package can perform the uptime check of several
         * monitors concurrently. Set this to a lower value if you're getting weird errors
         * running the uptime check.
         */
        'concurrent_checks' => 10,

        /*
         * The uptime check for a monitor will fail if the url does not respond after the
         * given number of seconds.
         */
        'timeout_per_site' => 10,

        /*
         * Because networks can be a bit unreliable the package can make three attempts
         * to connect to a server in one uptime check. You can specify the time in
         * milliseconds between each attempt.
         */
        'retry_connection_after_milliseconds' => 100,

        /*
         * Fire `Spatie\UptimeMonitor\Events\MonitorFailed` event only after
         * the given number of uptime checks have consecutively failed for a monitor.
         */
        'fire_monitor_failed_event_after_consecutive_failures' => 2,

        /*
         * When reaching out to sites this user agent will be used.
         */
        'user_agent' => 'spatie/laravel-uptime-monitor uptime checker',

        /*
         * When reaching out to the sites these headers will be added.
         */
        'additional_headers' => [],
    ],

    'certificate_check' => [

        /*
         * The `Spatie\UptimeMonitor\Events\SslExpiresSoon` event will fire
         * when a certificate is found whose expiration date is in
         * the next number of given days.
         */
        'fire_expiring_soon_event_if_certificate_expires_within_days' => 10,
    ],

    'expiration_check' => [

        /*
         * The `Spatie\UptimeMonitor\Events\DomainExpiresSoon` event will fire
         * when a domain is found whose expiration date is in
         * the next number of given days.
         */
        'fire_expiring_soon_event_if_domain_expires_within_days' => 30,

        /**
         * @todo Config option for custom whois servers
         * @see https://packagist.org/packages/io-developer/php-whois
         * @see https://github.com/io-developer/php-whois
         */
        'whois_servers' => [],
    ],

    /*
     * To add or modify behaviour to the Monitor model you can specify your
     * own model here. The only requirement is that it should extend
     * `Spatie\UptimeMonitor\Models\Monitor`.
     */
    'monitor_model' => Spatie\UptimeMonitor\Models\Monitor::class,
];
